### PR TITLE
Added new OpenVSP wrapper

### PR DESCRIPTION
**Pull request by anilyil**
## Purpose

This PR adds the new OpenVSP wrapper. The old one used file i/o. This one does everything in memory.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Testing

I have created a few tests, and will send them to @bbrelje. I will work with him to add these to pygeo tests.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Comment by anilyil** [comment]
The history has quite a few commits from my other work on DVGeometryMulti. I have removed that file for now. This PR only contains the work related to the new VSP wrapper.

**Comment by anilyil** [comment]
@joanibal do you remember the discussions we had on the methods `totalSensitivityProd` and `totalSensitivityTransProd`? It seems like these methods are inconsistent across various DVGeometry types. If you remember the main point of using `totalSensitivityTransProd` over `totalSensitivity`, please let me know and I can update those two methods in this PR. It seems like the openmdao wrapper in mphys uses these product methods, but it looks like `totalSensitivity` and `totalSensitivityTransProd` have identical functionalities in DVGeometry.py.

**Comment by joanibal** [comment]
@anil 

I still don't know, but here is the original slack question and resulting thread. 
image

image

**Review by joanibal — CHANGES_REQUESTED** [review]
Looks good, just a few minor questions/suggestions.

**Comment by anilyil** [comment]
Thanks for that comment @joanibal! Now I remember the deal with totalSensitivity and related methods. So there are three similar methods that I ran into:
1. `totalSensitivity`: This is the default MACH-API way of getting total derivatives w.r.t. design variables. Solvers (e.g. ADflow) call this internally with a `comm` object so that pygeo does the reduction itself. DVConstraint calls this without a comm object because those operations are serial, whereas parallel pointsets need to reduce the total sensitivities across the comm because the "reverse seeds" are distributed.
2. 'totalSensitivityProd': This is the unofficial way of doing forward AD with pygeo. I am not sure if we can consider this within the default MACH-API. ADflow does use this method in `computeJacobianVectorProductFwd`. However, DVCon does not use it (as far as I know). This method is provided with forward derivative seeds of design variables and it computes the fwd seeds of tracked points. Weirdly, this method also takes in a comm, but the comm object does not do anything. I think this method was written for cases where the DV fwd-seeds are set to the same values on all processors, and the code computes the fwd-seeds for the points tracked. Providing the comm is pointless because we cannot reduce this final fwd-seeds; the operation does not make sense physically, and the number of points can and does change across processors.
3. `totalSensitivityTransProd`: This is super similar to `totalSensitivity`, but it does not do a reduction in the comm. However, the method still takes in a comm. I am guessing this method was written for a non-official API to have an external framework (like OpenMDAO) to handle all the communication. But it is not written properly.

I believe my confusion started when I was first developing the new OpenVSP wrapper in this PR based on the older OpenVSP wrapper. In that version, the original VSP wrapper was designed to be mostly used with OpenMDAO. As a result, the OpenMDAO wrapper used methods numbered 2 and 3 in the list above. They look like they complement each other due to the name, but they dont exactly, and both have the un-used comm parameter.

Going forward, how I think we should fix this: We should get one version of the fwd and bwd code as standard, and remove all the extra versions. The optional comm argument to the bwd code is a good way to handle if reduction happens or not, so this function is versatile. And because `totalSensitivity` is used in a lot of places, I would take that as the default bwd method. Then, inside `totalSensitivityTransProd` method, we can just call `totalSensitivity` and that should just work. Finally, we can keep the forward code 'totalSensitivityProd', or create a new standard fwd code (we can call it 'totalSensitivityFwd' or sth similar)

Please feel free to correct me @eirikurj, especially if the methods I mentioned above are used with different purposes somewhere else in the code, especially in the aerostructural scripts.

Bottom line, we should have one forward and one backward sensitivity code, and we should remove all duplicates as @joanibal suggested.

File: pygeo/DVGeometryVSP.py [hunk 1]
```diff
@@ -411,91 +478,181 @@ def totalSensitivity(self, dIdpt, ptSetName, comm=None, config=None):
...
+        for i, dv in enumerate(self.DVs):
+            if dv in vec:
+                newvec[i] = vec[dv]
+
+        # we need to multiply with the surface jacobian
+        dPt = self.pointSets[ptSetName].jac.dot(newvec)
+
+        return dPt
+
+    def totalSensitivityTransProd(self, dIdpt, ptSetName, comm=None, config=None):
```
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:588]: see my comment below: https://github.com/mdolab/pygeo/pull/56#issuecomment-762425782

File: pygeo/DVGeometryVSP.py [hunk 2]
```diff
@@ -384,7 +452,6 @@ def getVarNames(self):
         """
         return list(self.DVs.keys())
```
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:453]: Good call. Will make that change in a bit

File: pygeo/DVGeometryVSP.py [hunk 3]
```diff
@@ -370,7 +431,14 @@ def pointSetUpToDate(self, ptSetName):
             return True
```
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:431]: That is just how its done across pygeo. See: https://github.com/mdolab/pygeo/blob/master/pygeo/DVGeometry.py#L1393-L1411 When a `pointSet` is added, an entry to the `self.updated` dictionary is also added. I believe that if check is to avoid the code raising a key error if the entry is not in the dictionary. I don't know if this regularly happens in the code; i.e. if some higher up code checks if a pointset is up to date even though the pointset is not tracked by that dvgeo.

File: pygeo/DVGeometryVSP.py [hunk 4]
```diff
@@ -610,378 +782,426 @@ def writePlot3D(self, fileName):
...
+        the VSP projection points, this is all we need to calculate. The input
+        pointSets is a list or dictionary of pointSets to calculate the jacobian for.
+
+        this routine runs in parallel, so it is important that any call leading to this
+        subroutine is performed synchronously among self.comm
+
+        VSP has a bug they refuse to fix. In a non-deterministic way, the spanwise u-v
+        mapping can differ from run to run. Seems like there are two modes. The differences
+        are small, but if we end up doing the FD with results from another processor
+        the difference is large enough to completely mess up the sensitivity. Due to this
```
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:924]: how about we track this in an issue #58. I will add a link to this issue in the code.

File: pygeo/DVGeometryVSP.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:453]: I ran into an issue in this process: The `self.DVs` is an ordered dictionary. The order is set by the order the DVs are added in the script. We have been assuming that this is the consistent across processors. May not be the best assumption but it has been true so far. 

This order determines the layout of the Jacobian, and the order of how we perturb the DVs. I cannot just sort the DVs at a point later, because we cannot guarantee the user will not add more DVs. 

So I can return `sorted( list( self.DVs.keys() ) )` here. But I will need to keep the ordering at https://github.com/mdolab/pygeo/blob/ec3b617f537891088af9f528877aa00d103c5923/pygeo/DVGeometryVSP.py#L938 

@joanibal, do you think there will be an issue with the ordering being different in the internal Jacobian vs the return value of `getVarNames`? To me, this is fine since this Jacobian is only used internally, and the sensitivities are communicated in dictionaries. But I may be missing a detail here.

[the remark this answers is not in the document]
Comment (joanibal) [reply pygeo/DVGeometryVSP.py:453]: I forgot that `self.DVs` was an ordered dict. Because it is ordered line 453 should be fine. 

I think the important thing is to make sure the keys are added to `self.DVs` in a consistent order. 
So maybe we should sort the keys from the user-supplied `dvDict` in `setDesignVars`.

[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:453]: Entries are added to self.DVs as the main runscript adds VSP variables to this DVGeo. In almost all scripts, this will be ordered the same on all procs, but I can see someone messing up the order.

However, I think with the general DVGeometry API, we do not really assume the user is ever done with adding additional DVs. i.e. the setDesignVars call does not need to indicate there won't be any more DVs. Although, the setDesignVars call kinda implies the addition of DVs are finished.

Finally, I think other DVGeos also rely on the top level script adding the DVs in a consistent order across processors. So I guess this is okay after all. What do you think?

[the remark this answers is not in the document]
Comment (joanibal) [reply pygeo/DVGeometryVSP.py:453]: OK. I see what your saying. 

I think it is fine after all

**Comment by anilyil** [comment]
@joanibal it seems like the only remaining issue is for us to decide on the totalSensitivity functions and what they do. I believe this is a pygeo-wide issue. I can create an issue and we can discuss it further there. I think we will also merge Ben's ESP wrapper at some point so we can fix all of those together consistently. 

Are you happy with the other parts of this PR? I would like to add a test case at least before we merge, but it's difficult bec. we will need to update our docker images with vsp. So maybe that can be left for another PR?

**Comment by bbrelje** [comment]
I plan on putting in some concerted effort on these reviews tomorrow during the day

**Review by bbrelje — APPROVED** [review]
This all looks nice and clean. Are we merging this before I add the unit tests (I'll add them this weekend in a separate PR) or do you want me to add the unit tests to this PR?

File: pygeo/DVGeometryVSP.py [hunk 5]
```diff
@@ -610,378 +782,427 @@ def writePlot3D(self, fileName):
...
+        # we now have all the point info on all procs.
+        tcomm += time.time() - t1
+
+        # We need to evaluate all the points on respective procs for FD computations
+
+        # allocate memory
+        pts0 = numpy.zeros((nptsg, 3))
+
+        # evaluate the points
+        for j in range(nptsg):
```
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:988]: Yeah I think so. Because we just use rigid links between the actual points and the projected points, we can just evaluate the sensitivities at the projected points. That is why those pts0 is the projected points, evaluated with whatever is in vsp currently. The pointset update is completely separate.

I believe I used the convention: pts for projected points, and "points" for the actual tracked coordinates. This Jacobian works on the pts. Does that make sense? Feel free to resolve this conversation if you are happy with my explanation, we can dig further if you want.

File: pygeo/DVGeometryVSP.py [hunk 6]
```diff
@@ -610,378 +782,427 @@ def writePlot3D(self, fileName):
...
+        t2 = time.time()
+        if rank == 0:
+            print("FD jacobian calcs with dvgeovsp took", (t2 - t1), "seconds in total")
+            print("updating the vsp model took", tvsp, "seconds")
+            print("evaluating the new points took", teval, "seconds")
+            print("communication took", tcomm, "seconds")
+
+        # set the update flags
+        for ptSet in self.pointSets:
+            self.updatedJac[ptSet] = True
```
[the remark this answers is not in the document]
Comment (anilyil) [reply pygeo/DVGeometryVSP.py:1103]: No particular reason. Its just easier to see that its done this way, vs looking at the other loops might be a bit more complicated. Its also separate from the computation etc. But yeah for efficiency, we could do them together, but the number of pointsets is often just a handful.

**Comment by anilyil** [comment]
@bbrelje do whatever is easier for you for the tests. You can push stuff to this branch or create a new PR with your branch.

**Review by joanibal — APPROVED** [review]

**Review by joanibal — APPROVED** [review]

**Comment by nwu63** [comment]
@bbrelje is the docker image with OpenVSP dependencies blocking this? If so I can try to work with Eirikur to get that sorted so we can merge this.

**Comment by bbrelje** [comment]
@nwu63 the ESP/VSP image is blocking right now, yes

**Comment by nwu63** [comment]
@bbrelje FYI the OpenVSP/ESP packages are now in deps

**Comment by bbrelje** [comment]
@anilyil please merge in my tests from my fork: https://github.com/bbrelje/pygeo-1/tree/vsp so we can get this PR closed out

**Comment by anilyil** [comment]
@bbrelje @nwu63 I merged Ben's branch here. Let me know if I need to do anything else.

**Comment by nwu63** [comment]
Oh no I guess we should merge the ESP one first so we get the STL for this? In the mean time should we add `skipTest` to these that depend on STL? Same with ESP tests too

**Comment by bbrelje** [comment]
@nwu63 we gotta draw the line somewhere with these skiptests. numpy-stl is required when installing the optional feature [testing] so I think it's reasonable for the tests to fail if numpy-stl is not present

but also not gonna die on this hill, so lmk how strongly you feel about that

**Review by bbrelje — DISMISSED** [review]

**Comment by nwu63** [comment]
> @nwu63 we gotta draw the line somewhere with these skiptests. numpy-stl is required when installing the optional feature [testing] so I think it's reasonable for the tests to fail if numpy-stl is not present
> 
> but also not gonna die on this hill, so lmk how strongly you feel about that

Yeah I guess this is required for testing, so what we have is fine. However, at this point why don't we just include STL as a required package (not under testing)? It's readily available from PyPI and installation is easy, so that might be easier than trying to manage multiple dependency sets. I'm also fine either way.

**Comment by nwu63** [comment]
@bbrelje See the Azure tests -- both of the Ubuntu images failed on one test, with atol of `4E-9` and `9E-7` against target atol of `1E-16`. I see there's an `rtol` of 1 but because the reference value is exactly 0, the `rtol` is pretty useless there. Do we want to adjust the tolerance? What's the expected accuracy here?

**Comment by marcomangano** [comment]
A single test is failing because the relative tolerance on a 0.0 ref value goes to `inf`
```
AssertionError: 
Not equal to tolerance rtol=1, atol=1e-06
Failed value for: %s_grad_error
Mismatched elements: 2 / 63 (3.17%)
Max absolute difference: 2.67608158e-06
Max relative difference: inf
 x: array([ 0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
        0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,
        0.000000e+00,  0.000000e+00,  0.000000e+00,  0.000000e+00,...
 y: array(0.)
```
Is there a way to go around this in `testflo`?

**Comment by nwu63** [comment]
Yeah the assert we use is a wrapper on [numpy.assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html), which basically checks for `atol` when the reference value for `rtol` is zero. This is exactly why we use both `atol` and `rtol` for testing, in case the reference value is very small.

**Review by bbrelje — APPROVED** [review]

**Comment by bbrelje** [comment]
@joanibal let's merge this - tests are passing now and opened an issue for the underlying problem

**Review by nwu63 — APPROVED** [review]
